activesupport/lib/active_support/core_ext/time/calculations.rb

```diff
@@ -105,6 +105,22 @@ def sec_fraction
     subsec
   end
 
+  unless Time.method_defined?(:floor)
+    def floor(precision = 0)
+      change(nsec: self.subsec.floor(precision) * 1_000_000_000)
+    end
+  end
+
+  unless Time.method_defined?(:ceil)
+    def ceil(precision = 0)
+      if precision == 0
+        change(sec: self.sec + 1, nsec: 0)
```

Comment:
Hmm, won't we need to chase this up? - i.e. 59 seconds, 59 minutes, etc.

Maybe this would be better:

``` ruby
def ceil(precision)
  change(nsec: 0) + subsec.ceil(precision)
end
```

Similarly, for `floor`:

``` ruby
def floor(precision)
  change(nsec: 0) + subsec.floor(precision)
end
```

Reply:
> Hmm, won't we need to chase this up? - i.e. 59 seconds, 59 minutes, etc.

`change` does chase up automatically and I've updated the test case to reflect that. However, I've updated the code to use your changes since it is more elegant with less lines of code.